internal/app/machined/pkg/system/runner/process/process.go

```diff
@@ -78,27 +88,90 @@ func (p *processRunner) Close() error {
 }
 
 type commandWrapper struct {
-	cmd              *exec.Cmd
+	launcher         *cap.Launcher
+	ctty             optional.Optional[int]
+	stdin            uintptr
+	stdout           uintptr
+	stderr           uintptr
 	afterStart       func()
 	afterTermination func() error
 }
 
+func dropCaps(droppedCapabilities []string, launcher *cap.Launcher) error {
+	droppedCaps := strings.Join(droppedCapabilities, ",")
+
+	prop, err := krnl.ReadParam(&kernel.Param{Key: "proc.sys.kernel.kexec_load_disabled"})
+	if v := strings.TrimSpace(string(prop)); err == nil && v != "0" {
+		log.Printf("kernel.kexec_load_disabled is %s, skipping dropping capabilities", v)
+	} else if droppedCaps != "" {
+		caps := strings.Split(droppedCaps, ",")
+		dropCaps := xslices.Map(caps, func(c string) cap.Value {
+			capability, capErr := cap.FromName(c)
+			if capErr != nil {
+				fmt.Printf("failed to parse capability: %s", capErr)
+			}
+
+			return capability
+		})
+
+		iab := cap.IABGetProc()
+		if err = iab.SetVector(cap.Bound, true, dropCaps...); err != nil {
+			return fmt.Errorf("failed to set capabilities: %w", err)
+		}
+
+		launcher.SetIAB(iab)
+	}
+
+	return nil
+}
+
+// This callback is run in the thread before executing child process.
+func beforeExecCallback(pa *syscall.ProcAttr, data interface{}) error {
+	wrapper, ok := data.(*commandWrapper)
+	if !ok {
+		return fmt.Errorf("failed to get command info")
+	}
+
+	ctty, cttySet := wrapper.ctty.Get()
+	if cttySet {
+		if pa.Sys == nil {
+			pa.Sys = &syscall.SysProcAttr{}
+		}
+
+		pa.Sys.Ctty = ctty
+		pa.Sys.Setsid = true
+		pa.Sys.Setctty = true
+	}
+
+	pa.Files = []uintptr{
+		wrapper.stdin,
+		wrapper.stdout,
+		wrapper.stderr,
+	}
+
+	// TODO: use pa.Sys.CgroupFD here when we can be sure clone3 is available
```

Comment:
let's absolutely do this

Reply:
So implement 2 different pathways? We use this code in container mode as well, and we cannot really expect containers to be ran on kernel 5.5+ which has clone3